File: dart/packages/aoc2017/lib/src/day01.dart
// https://adventofcode.com/2017/day/1

import 'package:aoc/aoc.dart';
import 'package:aoc/util/string.dart';
import 'package:collection/collection.dart';

main() => Day01().solve();

class Day01 extends AdventDay {
  Day01() : super(2017, 1);

  @override
  dynamic part1(String input) {
    final captcha = input.chars.map(int.parse).toList();
    final length = captcha.length;
    return captcha
        .whereIndexed((i, d) => captcha[i] == captcha[(i + 1) % length])
        .sum;
  }

  @override
  dynamic part2(String input) {
    final captcha = input.chars.map(int.parse).toList();
    final length = captcha.length;
    final halfway = length ~/ 2;
    return captcha
        .whereIndexed((i, d) => captcha[i] == captcha[(i + halfway) % length])
        .sum;
  }
}

File: clojure/src/aoc2015/day01.clj
;; https://adventofcode.com/2015/day/1
(ns aoc2015.day01
  (:require
   [aoc.day :as d]
   [aoc.util.collection :as c]
   [aoc.util.math :as m]))

(defn input [] (d/day-input 2015 1))

(def floor-delta {\( 1, \) -1})

(defn part1 [input]
  (m/sum (map floor-delta input)))

(defn part2 [input]
  (inc (c/index-of (reductions + (map floor-delta input)) -1)))
